### PR TITLE
[LV] Use IVUpdateMayOverflow to set HasNUW.

### PR DESCRIPTION
**Pull request by fhahn**
If IVUpdateMayOverflow is false, we proved that the induction increment cannot overflow in the vector loop. This allows setting NUW in some cases when folding the tail.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 1]
```diff
@@ -8943,9 +8943,12 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
...
   TailFoldingStyle Style = CM.getTailFoldingStyle(IVUpdateMayOverflow);
-  // When not folding the tail, we know that the induction increment will not
-  // overflow.
-  bool HasNUW = Style == TailFoldingStyle::None;
+  // Use NUW for the induction increment if we proved that it won't overflow in
+  // the vector loop or when not folding the tail. Then we know that the
+  // induction increment will not overflow as the vector trip count is >=
+  // increment and a multiple of the increment.
+  bool HasNUW = !IVUpdateMayOverflow || Style == TailFoldingStyle::None;
+  ;
```
Comment (artagnon) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:8951]: Stray ;?

Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Dropped, thanks!

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 2]
```diff
@@ -9014,9 +9014,11 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
 
   DebugLoc DL = getDebugLocFromInstOrOperands(Legal->getPrimaryInduction());
   TailFoldingStyle Style = CM.getTailFoldingStyle(IVUpdateMayOverflow);
-  // When not folding the tail, we know that the induction increment will not
-  // overflow.
-  bool HasNUW = Style == TailFoldingStyle::None;
+  // Use NUW for the induction increment if we proved that it won't overflow in
+  // the vector loop or when not folding the tail. Then we know that the
+  // induction increment will not overflow as the vector trip count is >=
+  // increment and a multiple of the increment.
```
Comment (artagnon) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:9020]: Is this correct: "the vector trip count >= increment and a multiple of the increment"? The trip count of the vectorized loop could be 1 for a VF of 4: won't the increment also be multiplied by 4 from the scalar version then? If the IV update doesn't overflow, doesn't it mean that trip count of vectorized loop * VF >= increment and a multiple of the increment? Maybe I'm getting confused with the "vector trip count" terminology.

**Review by artagnon — COMMENTED** [review]
I investigated the related code, and I think this is correct, modulo the confusing comment.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 3]
```diff
@@ -9014,9 +9014,11 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
...
+  bool HasNUW = !IVUpdateMayOverflow || Style == TailFoldingStyle::None;
```
Comment (artagnon) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:9021]: I think (although I'm not sure) you can drop the `Style == TailFoldingStyle::None`, as I think `!IVUpdateMayOverflow` subsumes it.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 2 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: The canonical induction goes from 0...Vector-trip-count, and is incremented by VF * UF in each iteration. The vector trip count is the original trip count adjusted to be a multiple of VF * UF.

I updated the comment to clarify that we mean the canonical induction increment. Does that make it clearer?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 3 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:9083]: It is still needed, as `IVUpdateMayOverflow` can only handle a subset of statically known cases; when not folding the tail, the runtime checks guarantee that there is no overflow.

**Review by artagnon — APPROVED** [review]
LGTM, thanks!

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 4]
```diff
@@ -9076,9 +9076,11 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
 
   DebugLoc DL = getDebugLocFromInstOrOperands(Legal->getPrimaryInduction());
   TailFoldingStyle Style = CM.getTailFoldingStyle(IVUpdateMayOverflow);
-  // When not folding the tail, we know that the induction increment will not
-  // overflow.
-  bool HasNUW = Style == TailFoldingStyle::None;
+  // Use NUW for the induction increment if we proved that it won't overflow in
+  // the vector loop or when not folding the tail. In the later case, we know
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:9080]: post-commit nit: "later" >> "latter"

**Pull request merged**